lint/mypy.bzl

```diff
@@ -0,0 +1,190 @@
+"""API for declaring a mypy lint aspect that visits py_library rules.
+
+Typical usage:
+
+```
+load("@aspect_rules_lint//lint:mypy.bzl", "mypy_aspect")
+
+mypy = mypy_aspect(
+    binary = "@@//:mypy",
+    configs = "@@//:pyproject.toml",
+)
+```
+
+This code inspired from https://github.com/bazel-contrib/bazel-mypy-integration
+with appreciation to [Jonathon Belotti](https://github.com/thundergolfer)
+and the other [contributors](https://github.com/bazel-contrib/bazel-mypy-integration/graphs/contributors)
+especially [David Zbarsky](https://github.com/dzbarsky)
+
+### TODO
+
+- Document --norun_validations
+- Support third-party definitions from stubs installed from type-* packages
+- Allow configured typeshed repo, e.g. args.add("--custom-typeshed-dir", "external/my_typeshed")
+- Set MYPYPATH if needed for cases where 1p sources don't resolve correctly to transitive .py sources: https://github.com/dzbarsky/mypy-bazel-experiments/blob/master/build/py/mypy.bzl#L163
+- Avoid invalidating caches whenever mypy.ini changes: https://github.com/dzbarsky/mypy-bazel-experiments/blob/master/build/py/mypy.bzl#L141-L149
+- Remote cache: bootstrap the stdlib since it will remain in cache, making other actions slightly faster: https://github.com/dzbarsky/mypy-bazel-experiments/blob/master/build/py/mypy.bzl#L451-L455
+- Later: Generate `.pyi` outputs: optimization to avoid as much invalidation when only implementation bodies change, at the cost of an extra mypy action.
+"""
+
+load("@bazel_skylib//lib:paths.bzl", "paths")
+
+_MNEMONIC = "mypy"
+MypyInfo = provider(doc = "Python typechecking data", fields = {
+    "transitive_cache_map": "depset: transitive --cache-map json files produced by deps",
+})
+
+def mypy_info(direct, deps):
+    return MypyInfo(
+        transitive_cache_map = depset(
+            direct = direct,
+            transitive = [d[MypyInfo].transitive_cache_map for d in deps if MypyInfo in d],
+        ),
+    )
+
+def _extract_transitive_inputs(deps):
+    result = []
+    for dep in deps:
+        # TODO: look through runfiles of 3p packages where .pyi files would appear in data.
```

Comment:
I remember we (me/Matt maybe?) were chatting a bit about surfacing these in the PyInfo provider, if that's not tractable maybe rules_py could hack something in? I suspect it would be both cleaner and better for perf than trying to find them here

Reply:
Yeah once this gets more mature I'll look at boiling up some of the parts into proposals on rules_python

Reply:
left a TODO for upstreaming something about stubs to rules_python.